### PR TITLE
Create structure, registr, receive message via email, reset password …

### PR DESCRIPTION
**Pull request by VitalyBashkiser**
…50%, access registration almost

### CODE REVIEW

File: api/database.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import os
+
+from sqlalchemy import create_engine
+from sqlalchemy.orm import DeclarativeBase
+from sqlalchemy.orm import sessionmaker, Session
+from dotenv import load_dotenv
+
+load_dotenv()
+
+DATABASE_URL = f"postgresql://{os.getenv('POSTGRES_USER')}:{os.getenv('POSTGRES_PASSWORD')}@{os.getenv('POSTGRES_HOST')}:{os.getenv('POSTGRES_PORT')}/{os.getenv('POSTGRES_DB')}"
```
Comment (SekachVitaliy) [inline api/database.py]: use the pydantic-settings for envs

File: apps/auth/api.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+async def register(user: user.UserCreate, db: Session = Depends(get_db)):
+    hashed_password = utils.hash_password(user.password)
+    db_user = User(email=user.email, hashed_password=hashed_password)
+    db.add(db_user)
+    db.commit()
+    db.refresh(db_user)
+
+    confirm_token = utils.create_confirmation_token(user.email)
+    host = "http://localhost:8000"
+    await email_service.reset_password(confirm_token, user.email, host)
```
Comment (SekachVitaliy) [inline apps/auth/api.py]: move logic to service. minimise code in routers

File: apps/auth/api.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+    return db_user
+
+
+@router.post("/login")
+def login(user: user.UserCreate, db: Session = Depends(get_db)):
+    db_user = db.query(User).filter(email=user.email).first()
+    if not db_user or not utils.verify_password(user.password, db_user.hashed_password):
+        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Invalid credentials")
```
Comment (SekachVitaliy) [inline apps/auth/api.py]: same and create a repository for database functions

File: apps/auth/api.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    return {"access_token": "fake-jwt-token", "token_type": "bearer"}
+
+
+@router.post("/password_reset")
+async def password_reset(user: user.ResetPassword, db: Session = Depends(get_db)):
+    email = user.email
+
+    db_user = db.query(User).filter(email=email).first()
+    if not db_user:
+        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="There is no such user.")
```
Comment (SekachVitaliy) [inline apps/auth/api.py]: same + add custom exceptions instead of HTTPException

File: apps/auth/api.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        )
+
+    db_user = db.query(User).filter(User=email).first()
+    if not db_user:
+        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="User not found.")
+
+    db_user.is_email_confirmed = True
+    db.add(db_user)
+    db.commit()
+
```
Comment (SekachVitaliy) [inline apps/auth/api.py]: same

File: apps/auth/email_service.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+import os
+from loguru import logger
+
+
+class EmailService:
+    def __init__(self):
+        self.smtp_server = "smtp.gmail.com"
+        self.smtp_port = 587
+        self.email = os.getenv("EMAIL_EMAIL")
+        self.password = os.getenv("EMAIL_PASSWORD")
```
Comment (SekachVitaliy) [inline apps/auth/email_service.py]: pydantic-settings

File: apps/auth/email_service.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        if reg:
+            logger.info(f"Confirm reg email to {email}, {reset_token=}")
+            subject = "Confirmation of Registration"
+            url = f"{host}/confirmation_of_registration/{reset_token}"
+            body = f"To confirm your registration, follow the link:\n{url}"
+        else:
+            logger.info(f"Send reset email to {email}, {reset_token=}")
+            subject = "Password Reset"
+            url = f"{host}/reset_password/{reset_token}"
+            body = f"To reset your password, follow the link:\n{url}"
```
Comment (SekachVitaliy) [inline apps/auth/email_service.py]: Do some refactoring . Reduce the code.

File: apps/schemas/user.py [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+    new_password: str
+
+
+class UserInDB(UserBase):
+    id: int
+    is_active: bool
+    is_email_confirmed: bool
+
+    class Config:
+        orm_mode = True
```
Comment (SekachVitaliy) [inline apps/schemas/user.py]: in pydantic v2 you need to use `model_config = ConfigDict(from_attributes=True)`

File: apps/user/api.py [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+
+router = APIRouter()
+
+
+@router.get("/{user_id}", response_model=user.UserInDB)
+def read_user(user_id: int, db: Session = Depends(get_db)):
+    db_user = db.query(User).filter(User=user_id).first()
+    if db_user is None:
+        raise HTTPException(status_code=404, detail="User not found")
+    return db_user
```
Comment (SekachVitaliy) [inline apps/user/api.py]: same. move to separate service

File: .idea/google-auth.iml [hunk 10]
```diff
@@ -4,7 +4,7 @@
     <content url="file://$MODULE_DIR$">
       <excludeFolder url="file://$MODULE_DIR$/.venv" />
     </content>
-    <orderEntry type="inheritedJdk" />
+    <orderEntry type="jdk" jdkName="Python 3.10 (google-auth)" jdkType="Python SDK" />
     <orderEntry type="sourceFolder" forTests="false" />
   </component>
 </module>
```
Comment (SekachVitaliy) [inline .idea/google-auth.iml]: move .idea to gitignore

File: .idea/misc.xml [hunk 11]
```diff
@@ -1,4 +1,7 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <project version="4">
-  <component name="ProjectRootManager" version="2" project-jdk-name="Python 3.11 (google-auth)" project-jdk-type="Python SDK" />
+  <component name="Black">
+    <option name="sdkName" value="Python 3.11 (google-auth)" />
+  </component>
+  <component name="ProjectRootManager" version="2" project-jdk-name="Python 3.10 (google-auth)" project-jdk-type="Python SDK" />
 </project>
```
Comment (SekachVitaliy) [inline .idea/misc.xml]: same. move .idea to gitignore

File: apps/auth/api.py [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+from apps import utils
+from apps.auth.email_service import email_service
+from apps.schemas import user
+from apps.user.models import User
+
+router = APIRouter()
+
+
+@router.post("/register", response_model=user.UserInDB)
+async def register(user: user.UserCreate, db: Session = Depends(get_db)):
```
Comment (SekachVitaliy) [inline apps/auth/api.py]: add docstrings to all fuctions and routers. for routers also add descriptions

File: api/database.py [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+
+engine = create_engine(DATABASE_URL)
+SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)
```
Comment (SekachVitaliy) [inline api/database.py]: You're using asynchronous functions, why not use an asynchronous connector to the db?

File: apps/auth/email_service.py [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+            with smtplib.SMTP("smtp.gmail.com", 587) as smtp:
+                smtp.ehlo()
+                smtp.starttls()
+                smtp.login(self.email, self.password)
+                smtp.sendmail(from_addr=self.email, to_addrs=recipients, msg=msg.as_string())
+            logger.info("Email sent successfully.")
+            return True
+        except Exception as e:
+            logger.exception(f"Failed to send email: {e}")
+            return False
```
Comment (SekachVitaliy) [inline apps/auth/email_service.py]: Extras. Make html templates for the emails and use them.

File: apps/user/models.py [hunk 15]
```diff
@@ -0,0 +1,13 @@
...
+
+
+class User(Base):
+    __tablename__ = "users"
+
+    id = Column(Integer, primary_key=True, index=True)
+    email = Column(String, unique=True, index=True)
+    hashed_password = Column(String)
+    is_active = Column(Boolean, default=False)
+    is_email_confirmed = Column(Boolean, default=False)
```
Comment (SekachVitaliy) [inline apps/user/models.py]: use SqlAlchemy ORM Mapped Class and alembic

**Review by SekachVitaliy — CHANGES_REQUESTED** [review]

File: src/api/routers/auth.py [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+    """
+    return await auth_service.change_email(uow, email_data, jwt_token)
+
+
+@router.post("/confirm_email_change", status_code=status.HTTP_200_OK)
+async def confirm_email_change(
+    data: OneTokenSchema,
+    uow: UOWDep,
+    auth_service: AuthService = Depends(),
+    jwt_token: Optional[str] = Depends(CheckHTTPBearer()),
```
Comment (SekachVitaliy) [inline src/api/routers/auth.py]: Don't use `Optional`. This is an obsolete method.
Use `str | None`
Change it all over the place.с

File: src/api/routers/auth.py [hunk 17]
```diff
@@ -0,0 +1,186 @@
...
+    prefix="/services",
+    tags=["Auth"],
+)
+
+
+@router.post("/register", status_code=status.HTTP_201_CREATED)
+async def register(
+    uow: UOWDep,
+    user: SchemeRegisterUser,
+    auth_service: AuthService = Depends(),
```
Comment (SekachVitaliy) [inline src/api/routers/auth.py]: Put it in dependencies.py

File: src/core/config/logger.py [hunk 18]
```diff
@@ -0,0 +1,5 @@
+import logging
+
+logging.basicConfig(level=logging.INFO, format="%(asctime)s - %(name)s - %(levelname)s - %(message)s")
+
+log = logging.getLogger(__name__)
```
Comment (SekachVitaliy) [inline src/core/config/logger.py]: Change it to `loguru`.

File: src/main.py [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+app = FastAPI(title="Authorisation via Email")
+
+app.add_exception_handler(UserNotFoundError, user_not_found_handler)
+app.add_exception_handler(UserAlreadyExistsError, user_already_exists_handler)
+app.add_exception_handler(InvalidCredentialsError, invalid_credentials_handler)
+app.add_exception_handler(EmailNotConfirmedError, email_not_confirmed_handler)
+app.add_exception_handler(InvalidTokenError, invalid_token_handler)
+app.add_exception_handler(PasswordResetError, handle_password_reset_error)
+app.add_exception_handler(EmailSendError, email_send_error_handler)
+app.add_exception_handler(UserNotAuthenticatedError, user_not_authenticated_handler)
```
Comment (SekachVitaliy) [inline src/main.py]: Put it in a function in a separate file, similar to routers.

File: src/services/auth_service.py [hunk 20]
```diff
@@ -0,0 +1,287 @@
...
+        if not current_user:
+            raise UserNotAuthenticatedError()
+
+        async with uow:
+            user = await uow.users.find_one_or_none(email=data.old_email)
+            if not user:
+                raise UserNotFoundError
+
+            confirm_token = auth_jwt.create_change_email_token(data.old_email, data.new_email)
+            host = settings.POSTGRES_HOST
```
Comment (SekachVitaliy) [inline src/services/auth_service.py]: Why POSTGRES?

File: src/services/users_service.py [hunk 21]
```diff
@@ -0,0 +1,34 @@
...
+            raise UserNotAuthenticatedError
+
+        async with uow:
+            user = await uow.users.find_one(id=user_id)
+            if user is None:
+                raise UserNotFoundError
+            return user
+
+
+user_service = UsersService()
```
Comment (SekachVitaliy) [inline src/services/users_service.py:32]: Why the initialisation?

File: src/utils/auth_jwt.py [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+    data = {"old_email": old_email, "new_email": new_email}
+    return create_access_token(data, expires_delta=timedelta(hours=1))
+
+
+def create_confirmation_token(email: str) -> str:
+    return create_access_token({"sub": email}, expires_delta=timedelta(hours=1))
+
+
+def create_reset_token(email: str) -> str:
+    return create_access_token({"sub": email}, expires_delta=timedelta(minutes=15))
```
Comment (SekachVitaliy) [inline src/utils/auth_jwt.py:109]: add docstrings

File: src/services/users_service.py (hunk 21 above)
Comment (VitalyBashkiser) [reply src/services/users_service.py:32]: This is with the analogy of auth, for further import of new routers and avoidance of circular import

File: requirements.txt
Comment (SekachVitaliy) [inline requirements.txt:1]: remove it)

File: pyproject.toml [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+readme = "README.md"
+
+[tool.poetry.dependencies]
+python = "^3.11"
+pycryptodome = "^3.21.0"
+python-jose = {extras = ["pycryptodome"], version = "^3.3.0"}
+aiohttp = "^3.10.10"
+alembic = "^1.14.0"
+black = "^24.10.0"
+isort = "^5.13.2"
```
Comment (SekachVitaliy) [inline pyproject.toml]: move to dev dependencies

File: src/main.py [hunk 24]
```diff
@@ -0,0 +1,19 @@
...
+
+setup_handlers(app)
+
+app.include_router(main_router)
+
+if __name__ == "__main__":
+    uvicorn.run(
+        "src.main:app",
+        host="0.0.0.0",
+        port=8090,
```
Comment (SekachVitaliy) [inline src/main.py]: make app_settings for host and port

File: src/services/auth_service.py [hunk 25]
```diff
@@ -0,0 +1,309 @@
...
+        """
+        user_dict = user.model_dump()
+        async with uow:
+            if await uow.users.find_one_or_none(email=user_dict["email"]):
+                raise UserAlreadyExistsError(user_dict["email"])
+
+            user_id = await uow.users.add_one(user_dict)
+
+            confirm_token = auth_jwt.create_confirmation_token(user_dict["email"])
+            host = "0.0.0.0"
```
Comment (SekachVitaliy) [inline src/services/auth_service.py]: create app_settings and use it to

File: src/services/auth_service.py [hunk 26]
```diff
@@ -0,0 +1,309 @@
...
+            Exception: If the user with the specified email is not found.
+        """
+        email = auth_jwt.verify_confirmation_token(token)
+        if email is None:
+            raise InvalidTokenError
+
+        async with uow:
+            user = await uow.users.find_one_or_none(email=email)
+
+            if user is None:
```
Comment (SekachVitaliy) [inline src/services/auth_service.py]: maybe `if not user`?

File: src/services/auth_service.py [hunk 27]
```diff
@@ -0,0 +1,309 @@
...
+        if not current_user:
+            raise UserNotAuthenticatedError
+
+        async with uow:
+            user = await uow.users.find_one_or_none(email=email)
+            if not user:
+                raise UserNotFoundError
+
+            reset_token = auth_jwt.create_reset_token(email)
+            host = "0.0.0.0"
```
Comment (SekachVitaliy) [inline src/services/auth_service.py]: same

File: src/services/auth_service.py [hunk 28]
```diff
@@ -0,0 +1,309 @@
...
+        if not current_user:
+            raise UserNotAuthenticatedError()
+
+        async with uow:
+            user = await uow.users.find_one_or_none(email=data.old_email)
+            if not user:
+                raise UserNotFoundError
+
+            confirm_token = auth_jwt.create_change_email_token(data.old_email, data.new_email)
+            host = "0.0.0.0"
```
Comment (SekachVitaliy) [inline src/services/auth_service.py]: same

**Review by SekachVitaliy — CHANGES_REQUESTED** [review]

**Pull request merged**